app/queues/dns/create-record-worker.server.ts

```diff
@@ -0,0 +1,130 @@
+import { UnrecoverableError, Worker } from 'bullmq';
+import type { Record } from '@prisma/client';
+import { RecordStatus } from '@prisma/client';
+
+import { redis } from '~/lib/redis.server';
+import { prisma } from '~/db.server';
+import logger from '~/lib/logger.server';
+import { setMonthsFromNow } from '~/lib/domains.server';
+import { createRecord, getChangeStatus } from '~/lib/dns.server';
+
+export const addDbRecordQueueName = 'add-db-record';
+export const createDnsRecordQueueName = 'create-dns-record';
+export const checkDnsStatusQueueName = 'check-dns-status';
+export const syncDbStatusQueueName = 'sync-db-status';
+
+export interface JobRecord {
+  username: Record['username'];
+  type: Record['type'];
+  name: Record['name'];
+  value: Record['value'];
+}
+
+export const addDbRecordWorker = new Worker<JobRecord>(
+  addDbRecordQueueName,
+  async (job) => {
+    logger.info('Creating a record in DB');
+    const { username, type, name, value } = job.data;
+
+    const existingRecords = await prisma.record.count({
+      where: {
+        type,
+        name,
+        value,
+      },
+    });
+
+    if (existingRecords) {
+      throw new Error('Record already exists');
+    }
+
+    try {
+      const result = await prisma.record.create({
+        data: {
+          username,
+          type,
+          name,
+          value,
+          expiresAt: setMonthsFromNow(6),
+          status: RecordStatus.pending,
+        },
+      });
+      return result.id;
+    } catch (error) {
+      logger.warn('Could not create a record in DB', error);
+      throw error;
+    }
+  },
+  { connection: redis }
+);
+
+export const createDnsRecordWorker = new Worker<JobRecord>(
+  createDnsRecordQueueName,
+  async (job) => {
+    logger.info('Creating a record in Route 53');
+    const { username, type, name, value } = job.data;
+
+    try {
+      return createRecord(username, type, name.toLowerCase(), value);
+    } catch (error) {
+      logger.warn('Could not create a record in Route53', error);
+      throw error;
+    }
+  },
+  { connection: redis }
+);
+
+export const checkDnsStatusWorker = new Worker(
+  checkDnsStatusQueueName,
+  async (job) => {
+    logger.info('Checking Route53 readiness');
+
+    const values: { [jobKey: string]: string } = await job.getChildrenValues();
+    const key = Object.keys(values)[0];
+    const changeId = values[key];
+
+    const status = await getChangeStatus(changeId);
+
+    if (status === 'PENDING') {
+      throw new Error('Change status is still pending');
+    }
+    return status;
+  },
+  { connection: redis }
+);
+
+export const syncDbStatusWorker = new Worker(
+  syncDbStatusQueueName,
+  async (job) => {
+    logger.info('Sync DB status with DNS');
+
+    const childJobValues = await job.getChildrenValues();
+    if (Object.keys(childJobValues).length !== 2) {
```

Comment:
You use this repeatedly, you might as well cache it and reuse:

```ts
const values = Object.keys(childJobValues);
```

Reply:
I'm not sure how cache should be done. Can I do with `global`?

Reply:
I totally misunderstood the comment. Sorry for the wrong question.